### PR TITLE
Sweetandsour2

### PR DESCRIPTION
**Pull request by bthaeler**
This is a new hook system that replaces the embedded perf code. This will be the foundation for a new debugger interface.

### CODE REVIEW

**Comment by caridy** [comment]
@bthaeler a couple of questions:
- did you validate the unit tests? I remember when I put in place the perf there were few issues with the tests due to the fact that every script is using Y.mojito.hooks, which is undefined if it is not part of the testsuite mocks.
- is the current output format of the perf still preserved? if not, we need to verify with @drewfish about this impact of that since we use it in the current profiler and potentially in our CI to monitor performance, but I'm not sure about that last part.

**Comment by caridy** [comment]
According to @drewfish we are not using perf and profiler at all in our CI, so, feel free to break it, and even remove it in favor of the new profiler.

So, the only thing pending is the tests to pass.

**Comment by bthaeler** [comment]
The existing perf markers are re-implemented using the hook system. So they should continue to work just fine.

It wasn't very clear to me how to run the unit tests. So I didn't. Looking at the Travis build output on github, it looks like some of the tests do complain. I can try and fix those. Is there a good current document explaining how to setup a test system for running the unit tests?

**Comment by caridy** [comment]
Few pointers:
- everyone who uses hooks should requires mojito-hooks.
- mock it in https://github.com/yahoo/mojito/blob/develop/tests/base/mojito-test.js to do nothing when running tests.
- to run tests, check:
  cd tests/ 
  ./run.js help
- more info about tests here: https://github.com/yahoo/mojito/tree/develop/tests

**Comment by bthaeler** [comment]
I added some missing mojito-hooks dependency to places, and I added a stuff for the hook system to the test base. I think the unit tests are fine now.

File: tests/base/mojito-test.js [hunk 1]
```diff
@@ -26,6 +26,12 @@ YUI.add('mojito', function(Y, NAME) {
     Y.namespace('mojito.addons');
     Y.namespace('mojito.addons.ac');
     Y.namespace('mojito.addons.viewEngines');
+    // HookSystem::StartBlock
+    Y.mojito.hooks = {
+        hook: function () {},
```
Comment (caridy) [inline tests/base/mojito-test.js:31]: I think `enableHookGroup` is also missing from this. Even though only perf is calling it conditionally, I think we will have more scripts calling it, in which case we need a mock for testing.

File: lib/mojito.js [hunk 2]
```diff
@@ -281,6 +281,13 @@ MojitoServer.prototype._configureAppInstance = function(app, options) {
         outputHandler.setLogger({
             log: Y.log
         });
+        // HookSystem::StartBlock
+        if (!req.hook) {
+            req.hook = null;
```
Comment (caridy) [inline lib/mojito.js:286]: What is the purpose of setting it to `null` when it is already a falsy value?

File: lib/mojito.js [hunk 3]
```diff
@@ -281,6 +281,13 @@ MojitoServer.prototype._configureAppInstance = function(app, options) {
...
+        }
+        outputHandler.hook = req.hook;
+        command.instance.hook = req.hook;
```
Comment (caridy) [inline lib/mojito.js]: Any particular reason for adding hook into the `instance` instead of `command`. In theory, an instance's action can only be executed if there is a command to be dispatched. In my mind, `hook` is more tied to `command` than it is to `instance`. In any case, you should have access to `command` from everywhere during the dispatch process, just like instance. In the server, this is just a technicallity, but in the client, that instance is persistent, and command is created every time an action is invoked.

File: lib/app/autoload/mojito-client.client.js [hunk 4]
```diff
@@ -629,6 +642,12 @@ YUI.add('mojito-client', function(Y, NAME) {
 
             outputHandler = new Y.mojito.OutputHandler(viewId, cb, this);
 
+            // HookSystem::StartBlock
+            if (Y.mojito.hooks) {
+                outputHandler.hook = Y.mojito.hooks.client_ctx;
```
Comment (caridy) [inline lib/app/autoload/mojito-client.client.js]: `client_ctx` sounds weird here, specially when you assign it to `outputHandler.hook`. Is client_ctx  the persistent hook in the client side? If yes, can we rename it? We don't want to keep overloading the `context` term in mojito, we already use it for many things.

File: lib/app/autoload/hooks.common.js [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+     * Y.mojito.hooks.registerHook('test_tool, 'test_hook', function (reg, data) {});
+     * </pre>
+     */
+    Y.mojito.hooks.registerHook = function (tool, hook, cb) {
+        if (!map_hook_tool[hook]) {
+            map_hook_tool[hook] = {};
+        }
+        map_hook_tool[hook][tool] = cb;
+
+        Y.mojito.hooks.hook = Y.mojito.hooks.real_hook;
```
Comment (caridy) [inline lib/app/autoload/hooks.common.js:116]: Why is this happening here? This sounds like a leak-prompt structure, because any request can potentially change this. If the hook is something created per request, then it should be an inmutable object.

File: lib/app/autoload/hooks.common.js [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+     * Example:
+     * <pre>
+     * Y.mojito.hooks.enableHookGroup(req, 'test_tool');
+     * </pre>
+     */
+    Y.mojito.hooks.enableHookGroup = function (req, tool) {
+        var ret = {};
+        // Y.log("enable tool: " + tool);
+
+        // if called in client, use single global context.
```
Comment (caridy) [inline lib/app/autoload/hooks.common.js:138]: In my experience, trying to accomodate a component to be `common` when in reality the behavior of the client and server are very different will bring problems. I wonder if we should split this into server and client.

File: lib/app/autoload/hooks.common.js [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+     * <pre>
+     * Y.mojito.hooks.hook('test_hook', ctx, ...);
+     * </pre>
+     */
+    Y.mojito.hooks.hook = function () {
+        // Y.log("hooks disabled");
+    };
+
+    Y.mojito.hooks.real_hook = function (hook, ctx) {
+        // Y.log("in hook handler: " + hook);
```
Comment (caridy) [inline lib/app/autoload/hooks.common.js]: Please, preserve all Y.log(msg, 'info or debug or error', NAME), instead of commenting them.

File: lib/app/autoload/hooks.common.js [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+/**
+ * The hook system provides a way for application or add on developers to access the inner working of mojito.
+ * There are two steps to using an hook. First, an addon needs to register an interest in a hook by calling
+ * registerHook with the name of the hook and a callback function. The second step involves enabling a hook
+ * your addon to recieve hooks. You need to call the enableHookGroup with the name of your addon.
+ *
+ * List of hooks
+ *
+ * <ul>
+ * <li>Y.mojito.hooks.hook('adapterBuffer', hook_ctx, 'start', this);
```
Comment (caridy) [inline lib/app/autoload/hooks.common.js:20]: Why do we have these docs here? this is the hook utility, it will be difficult to keep this in sync with all the hooks defined across the mojito or the app. We can create a tool to generate such as list, or we can document them in a form of YUI events on site, rather than here.

File: lib/app/addons/view-engines/hb.server.js [hunk 9]
```diff
@@ -39,26 +39,32 @@ YUI.add('mojito-hb', function(Y, NAME) {
          */
         render: function (data, mojitType, tmpl, adapter, meta, more) {
             var cacheTemplates = meta && meta.view && meta.view.cacheTemplates,
-                perf = Y.mojito.perf.timeline('mojito', 'hb:render',
-                    'time to render a template', tmpl),
-                handler = function (err, obj) {
-                    var output;
+                handler;
```
Comment (caridy) [inline lib/app/addons/view-engines/hb.server.js]: Any particular reason for moving handler's function definition down rather than in the var statement?

File: lib/app/addons/ac/composite.common.js [hunk 10]
```diff
@@ -204,8 +209,10 @@ callback({
             var ac = this.ac,
                 buffer = {},
                 content = {},
-                meta = {},
-                perf;
+                // HookSystem::StartBlock
+                self = this,
```
Comment (caridy) [inline lib/app/addons/ac/composite.common.js]: my convention, we use `my = this` instead of `self = this`. Don't ask my why :)

And by the way, don't need to define the comments for that line, it is ok to have `my` defined even when u don't use it later on.

File: lib/app/addons/ac/composite.common.js [hunk 11]
```diff
@@ -14,19 +14,24 @@
...
 
-    function AdapterBuffer(buffer, id, callback) {
+    function AdapterBuffer(buffer, id, callback, hook_ctx) {
         this.buffer = buffer;
         this.id = id;
         this.callback = callback;
         this.__meta = [];
-        this.__perf = Y.mojito.perf.timeline(NAME, 'child', 'the whole child', id);
+        // HookSystem::StartBlock
+        this.__hook_ctx = hook_ctx;
```
Comment (caridy) [inline lib/app/addons/ac/composite.common.js]: Can you elaborate more about this `__hook_ctx` very private reference?

File: lib/app/addons/ac/composite.common.js [hunk 12]
```diff
@@ -14,19 +14,24 @@
  */
 YUI.add('mojito-composite-addon', function(Y, NAME) {
 
-    function AdapterBuffer(buffer, id, callback) {
+    function AdapterBuffer(buffer, id, callback, hook_ctx) {
```
Comment (caridy) [inline lib/app/addons/ac/composite.common.js]: This is the only part that I really don't like. Extending the signature of the class to add the hook reference after the callback is not pretty. Let's try to find another way to keep that reference around.

File: lib/app/addons/ac/composite.common.js [hunk 13]
```diff
@@ -335,7 +346,7 @@ callback({
                     };
 
                     childAdapter = new AdapterBuffer(buffer, childName,
-                            callback);
+                            callback, this.adapter.hook);
```
Comment (caridy) [inline lib/app/addons/ac/composite.common.js]: same as above, we should not do this here. Let's find another way for the adapter to pull a reference of the current hook.

**Comment by caridy** [comment]
@bthaeler that's all for now. Let's work together to accomodate those details so we can finally merge it.

File: lib/mojito.js (hunk 2 above)
Comment (bthaeler) [inline lib/mojito.js:286]: From my, and your performance tests, it is more efficient to test if a field is null vs undefined. IE 'if (x.y)' is more efficient if x.y = null vs x.y is just undefined.

I am just trying to make the hook system as low impact as possible.

File: lib/app/autoload/mojito-client.client.js (hunk 4 above)
Comment (bthaeler) [inline lib/app/autoload/mojito-client.client.js]: Originally this was for the clients. But this also works for people who want to enable a hook global, without associating this with a request. So how about 'global_hooks'.

Each callback will have its own unique this state per tool. This is stored in this object. So do we want 'global_hooks' or 'global_hooks_ctx'?

File: lib/app/autoload/hooks.common.js (hunk 5 above)
Comment (bthaeler) [inline lib/app/autoload/hooks.common.js:116]: Not sure what your concern is here. Hooks are statically defined. Groups of hooks are enabled on a per request basis. What hooks people are interested in is static.

File: lib/app/autoload/hooks.common.js (hunk 6 above)
Comment (bthaeler) [inline lib/app/autoload/hooks.common.js:138]: If I rename this to 'global_hooks' I think this resolved this.

File: lib/app/addons/view-engines/hb.server.js (hunk 9 above)
Comment (bthaeler) [inline lib/app/addons/view-engines/hb.server.js]: No, i will fix.

File: tests/base/mojito-test.js (hunk 1 above)
Comment (caridy) [inline tests/base/mojito-test.js:31]: Any update on this one?

File: lib/mojito.js (hunk 2 above)
Comment (caridy) [inline lib/mojito.js:286]: ok, we can revisit this later on after the first merge.

File: lib/app/autoload/hooks.common.js (hunk 8 above)
Comment (caridy) [inline lib/app/autoload/hooks.common.js:20]: any update on this one?

File: lib/app/autoload/hooks.common.js (hunk 5 above)
Comment (caridy) [inline lib/app/autoload/hooks.common.js:116]: My question is why hook gets real_hook whenever a new hook is register? is this something to have "hook" defined only if there is at least one hook registered? I'm just curious about this statement.

File: tests/base/mojito-test.js (hunk 1 above)
Comment (bthaeler) [inline tests/base/mojito-test.js:31]: yes, I added a stub function.

```
    enableHookGroup: function () {}
```

File: lib/app/autoload/hooks.common.js (hunk 8 above)
Comment (bthaeler) [inline lib/app/autoload/hooks.common.js:20]: Sounds like you are proposing two options. One is to create a script that runs at build time to generate this list, and the second one is using some kind of YUI event to do this? I don't now what would be better. Is this something we want to solve now or latter? If you feel strongly about this, I can try and figure something out.

File: lib/app/autoload/hooks.common.js (hunk 5 above)
Comment (bthaeler) [inline lib/app/autoload/hooks.common.js:116]: Yes, the idea is that if you are not using any code that needs a hook, the hook system defaults to an empty function.

Comment (caridy) [inline lib/app/autoload/hooks.common.js:116]: ok

**Comment by caridy** [comment]
+1

@bthaeler I think we are ready to merge it. Since we have a release coming up on monday, I will hold this until after that. 

/cc @jenny

**Comment by bthaeler** [comment]
Any idea when you will do the merge for this pull request?

**Comment by caridy** [comment]
today or tomorrow! 0.5.2 is out!
